source/slang/slang-ir-spirv-legalize.cpp

```diff
@@ -1167,6 +1167,166 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
         addUsersToWorkList(newStore);
     }
 
+    void processNonUniformResourceIndex(IRInst* nonUniformResourceIndexInst)
+    {
+        // implement the translation to spirv by waking up the use-def chain
```

Comment:
I think the logic here should be: if we see `gep(pArray, nonUniformRes(x))`, then replace it with
`nonUniformRes(gep(pArray, x))` then add this new nonUniformRes inst to work list.
If we see `instCast(nonUniformRes(x))`, do the same and replace it with `nonUniformRes(intCast(x))`.

If we see `nonUniformRes(load(x))` or any other unknown opcodes, stop.

Then after that, replace all `nonUniformRes(x)` with `x`, and add a nonuniform res decoration to `x`.

Peudo code:
```
workList.add(nonUniformResourceIndexInst);
for (Index i = 0; i < workList.getCount(); i++)
{
      traverseUsers(inst)
      {
            auto user = use->getUser();
            IRBuilder builder(user);
            builder.setInsertBefore(user);
            IRInst* newUser = nullptr;
            switch(user->getOp())
            {
            case IntCast:
                  newUser = builder.emitIntCast(user->getFullType(), inst->getOperand(0);
                  break;
            case GEP:
                    if (use is operand 1)
                        newUser = builder.emitGEP(user->getFullType(), inst->getOperand(0);
                   break;
            case Load:
                    newUser = builder.emitLoad(user->getFullType(), inst->getOperand(0);
                    break;
            }
           if (!newUser) continue;
           auto nonUniformUser = builder.emitNonUniformResourceIndex(newUser);
           user->replaceAllUsesWith(nonUniformUser);
           switch(user->getOp())
           {
           case IntCast:
           case GEP:
                  workList.add(nonUniformUser);
            }
      }
}

// Now we have "upstream"ed all `NonUniformResourceIndex` insts, 
// the next step is to add a decoration the operands of all nonuniformresourceindex insts
for (auto nonUniformResInst : workList)
      auto operand =  nonUniformResInst->getOperand(0);
      if (isResourceType(operand->getDataType()) || operand's type is a pointer to resource type)
          builder.addNonUnifomResourceDecoration(operand);
      nonUniformResInst->replaceUsesWith(operand);

```

Reply:
Seems like this is going in the right direction. Instead of using a for loop, I used a recursive visitor 
```
 std::function<void(IRInst*)> visit = [&](IRInst* inst)
 {
     traverseUses(inst, [&](IRUse* use)
     {
         auto user = use->getUser();
         IRBuilder builder(user);
         builder.setInsertBefore(user);
         IRInst* newUser = nullptr;
         switch (user->getOp())
         {
         case kIROp_IntCast:
             newUser = builder.emitCast(user->getFullType(), inst->getOperand(0));
             break;
         case kIROp_GetElementPtr:
             newUser = builder.emitElementAddress(user->getFullType(), user->getOperand(0), inst->getOperand(0));
             break;
         case kIROp_Load:
             newUser = builder.emitLoad(user->getFullType(), inst->getOperand(0));
             break;
         default:
             break;
         };
         if (newUser)
         {
             auto nonUniformUser = builder.emitNonUniformResourceIndex(newUser->getFullType(), newUser);
             user->replaceUsesWith(nonUniformUser);
         }
     });
 };
 visit(nonUniformResourceIndexInst);
```
The logic is almost the same, except for the fact that we don't have a worklist at hand to deal with the next steps.
What is interesting in this approach is that the, processWorkList() function that invokes this code within `processNonUniformResourceIndex`, actually keeps track of all the instructions, and the first time it calls the function it processes the GEP, and the next time it processes the Load. It is intelligent to keep checking the IR until all `NonUniform`s are processed.

After applying the first part of the transform, this is what I am getting. 
```
	{1710}	let  %21	: Int	= nonUniformResourceIndex(0 : Int)
	{1767}	let  %22	: TextureType(Float, TextureShape1DType, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, %9)	= getElement(%textures, %21)
	{2714}	let  %23	: Vec(Int, 2 : Int)	= makeVector(0 : Int, 0 : Int)
	{2688}	SPIRVAsm %24	: Float
	{
		{2692}	SPIRVAsmInst(95 : UInt, __sampledType(Float), %"sampled", getElement(%textures, %21), makeVector(0 : Int, 0 : Int), 2 : UInt, 0 : Int)
		{2694}	SPIRVAsmInst(__truncate, Float, result, __sampledType(Float), %"sampled")
	}
```
to
```
	{1710}	let  %21	: Int	= nonUniformResourceIndex(0 : Int)
	{2775}	let  %22	: Ptr(TextureType(Float, TextureShape1DType, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, %18), 0 : Int)	= getElementPtr(%textures, 0 : Int)
	{2776}	let  %23	: Ptr(TextureType(Float, TextureShape1DType, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, %18), 0 : Int)	= nonUniformResourceIndex(%22)
	{2780}	let  %24	: Ptr(TextureType(Float, TextureShape1DType, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, %18), 0 : Int)	= nonUniformResourceIndex(%22)
	{2781}	let  %25	: TextureType(Float, TextureShape1DType, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, %18)	= load(%22)
	{2782}	let  %26	: TextureType(Float, TextureShape1DType, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, %18)	= nonUniformResourceIndex(%25)
	{2790}	let  %27	: Vec(Int, 2 : Int)	= globalValueRef(%28)
	{2688}	SPIRVAsm %29	: Float
	{
		{2692}	SPIRVAsmInst(95 : UInt, __sampledType(Float), %"sampled", nonUniformResourceIndex(%25), globalValueRef(%28), 2 : UInt, 0 : Int)
		{2694}	SPIRVAsmInst(__truncate, Float, result, __sampledType(Float), %"sampled")
	}
```

If this is the case, can I perform the inst -> decoration conversion in the spirv emitter?

Reply:
The second step shall be done inside the function processing function.

The philosophy is never do any actual IR transformation during emit. Emit should just be straight forward conversion from slang ir to spirv.

Reply:
The work list mechanism is to speed up the process so we don't need to rediscover all nonUniformInsts. We also don't want to involve any recursion because it is slow and risks overflowing the stack.

Reply:
understood!

Reply:
This is the final IR I'm seeing after all the fixups:
```
	[{2796}	NonUniformResource]
	[{2784}	NonUniformResource]
	{2775}	let  %21	: Ptr(TextureType(Float, TextureShape1DType, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, %18), 0 : Int)	= getElementPtr(%textures, 0 : Int)
	[{2798}	NonUniformResource]
	{2778}	let  %22	: TextureType(Float, TextureShape1DType, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, 0 : Int, %18)	= load(%21)
	{2806}	let  %23	: Vec(Int, 2 : Int)	= globalValueRef(%24)
	{2688}	SPIRVAsm %25	: Float
	{
		{2692}	SPIRVAsmInst(95 : UInt, __sampledType(Float), %"sampled", load(%21), globalValueRef(%24), 2 : UInt, 0 : Int)
		{2694}	SPIRVAsmInst(__truncate, Float, result, __sampledType(Float), %"sampled")
	}
```
 Does it look like its in the right step.